source/modulo_components/modulo_components/lifecycle_component.py

```diff
@@ -0,0 +1,240 @@
+        self._change_state_srv = self.create_service(ChangeState, '~/change_state', self.__change_state)
+
+        self.add_predicate("is_unconfigured", lambda: self.__state == State.PRIMARY_STATE_UNCONFIGURED)
+        self.add_predicate("is_inactive", lambda: self.__state == State.PRIMARY_STATE_INACTIVE)
+        self.add_predicate("is_active", lambda: self.__state == State.PRIMARY_STATE_ACTIVE)
+        self.add_predicate("is_finalized", lambda: self.__state == State.PRIMARY_STATE_FINALIZED)
+
+    def __change_state(self, request: ChangeState.Request, response: ChangeState.Response) -> ChangeState.Response:
+        """
+        Change state service callback which calls the requested transition method.
+
+        :param request: The lifecycle_msgs.msg.ChangeState.Request message
+        :param response: The lifecycle_msgs.msg.ChangeState.Response message
+        :return: The response message
+        """
+        self.get_logger().debug(f'Change state service called with request {request}')
+        if request.transition.id == Transition.TRANSITION_CONFIGURE:
+            response = self.__on_configure(request, response)
+        elif request.transition.id == Transition.TRANSITION_ACTIVATE:
+            response = self.__on_activate(request, response)
+        elif request.transition.id == Transition.TRANSITION_DEACTIVATE:
+            response = self.__on_deactivate(request, response)
+        # TODO the other transitions
+        else:
+            self.get_logger().error(f'Unsupported state transition! {request}')
+            response.success = False
+        return response
+
+    def __on_configure(self, request: ChangeState.Request, response: ChangeState.Response) -> ChangeState.Response:
+        """
+        Configure service callback. Simply call the handle_configure function.
+
+        :param request: Request to change the state to inactive
+        :param response: The response object from the service call
+        :return: True if the component configured successfully
+        """
+        if not self.__state == State.PRIMARY_STATE_UNCONFIGURED:
+            response.success = False
+            return response
+
+        response.success = self.__handle_configure()
+        if response.success:
+            self.__state = State.PRIMARY_STATE_INACTIVE
+        return response
+
+    def __handle_configure(self) -> bool:
+        """
+        Handle the configure transition by related transition steps and invoking the user callback.
+
+        :return: True if the transition was successful
+        """
+        return self.__configure_outputs() and self.on_configure_callbacck()
+
+    def on_configure_callbacck(self) -> bool:
+        """
+        Function called from the configure transition service callback. To be redefined in derived classes.
+
+        :return: True if the component configured successfully
+        """
+        return True
+
+    def __configure_outputs(self) -> bool:
```

Comment:
It's a minor thing but I would move this to a different part of the file, basically grouping all the __on_xyz, __handle_xyz, on_xyz_callback together without interruption